bugsnag-android-core/src/main/java/com/bugsnag/android/FileStore.java

```diff
@@ -45,32 +42,30 @@
     private final Logger logger;
     private final EventStore.Delegate delegate;
 
-    FileStore(@NonNull Context appContext, String folder,
-              int maxStoreCount, Comparator<File> comparator, Logger logger, Delegate delegate) {
+    FileStore(@NonNull File storageDir,
+              int maxStoreCount,
+              Comparator<File> comparator,
+              Logger logger,
+              Delegate delegate) {
         this.maxStoreCount = maxStoreCount;
         this.comparator = comparator;
         this.logger = logger;
         this.delegate = delegate;
+        this.storageDir = storageDir;
 
-        String path;
         try {
-            path = appContext.getCacheDir().getAbsolutePath() + folder;
-
-            File outFile = new File(path);
-            outFile.mkdirs();
-            if (!outFile.exists()) {
-                this.logger.w("Could not prepare file storage directory");
-                path = null;
+            storageDir.mkdirs();
+            if (!storageDir.isDirectory() || !storageDir.canWrite()) {
```

Comment:
`storageDir` can no longer be `null`, so some of the guards in `FileStore` and `EventStore` need updating. 

Perhaps we should set `storageDir` to `null` here in the error case and guard against `null` everywhere? Either way, make sure all the operations on `storageDir` are suitably guarded/caught to make to avoid an exception.

Reply:
Updated to remove null guards. I think it's preferable to check whether the directory exists when writing to it, as some places were assuming that `storageDir` was non-null anyway - e.g. `SessionStore#getFilename()`.

Reply:
Can I propose that we go the other direction and set `storageDir` to `null` in the constructor if the value is not suitable (i.e. not a writable directory). It feels like allowing things to proceed with just a log message leaves the processing open to issues. For example, if the directory is not writable then it looks like the `write` and `enqueueContentForDelivery` methods would cause an internal report for something that is a configuration issue.

However, if all the file IO was skipped if there was no storageDir, we can ensure that no processing happens if validation in the constructor fails (rather than distributing the validation throughout the code).

To make this better encapsulated, I believe we could move the directory part of the filename out of the duty of `getFilename` for the Event- and SessionStores, so that it literally just returns the file name. The directory can then be prepended by FileStore (having performed the necessary null check).